Implement groups on the canvas for organizing pods.

chimmyten opened the pull request:
Users can add pods to groups, and the groups should expand or shrink automatically.

chimmyten committed:
render static group node

chimmyten committed:
add scope option to context menu

chimmyten committed:
change scope to group

chimmyten commented:
<img width="966" alt="groupnode-static-render" src="https://github.com/JupyPod/JupyPod/assets/105993789/be2a9d80-6fde-4e5a-b99b-55d61c8dbee7">

forrestbao commented:
Thanks @chimmyten Please document how groups are stored in NBFormat.

chimmyten commented:
Sure I'll do that. Should I implement saving the groups in this pr as well? Or should that be a separate one?

forrestbao commented:
> Sure I'll do that. Should I implement saving the groups in this pr as well? Or should that be a separate one?

Yeah, save all information on a canvas to a JSON file, as well as be able to load a canvas from a JSON file.

chimmyten commented:
Oh, I meant should I make a separate PR for that feature?

forrestbao commented:
> Oh, I meant should I make a separate PR for that feature?

sounds a good idea. Keep PRs modualarized.

chimmyten committed:
implement adding group to canvas via context menu

chimmyten committed:
add delete button to groups

chimmyten commented:
https://github.com/JupyPod/JupyPod/assets/105993789/3fc5da61-0183-459f-a9f0-c14e7f136bc3

chimmyten committed:
highlight scope when node is hovered over

chimmyten commented:
https://github.com/JupyPod/JupyPod/assets/105993789/b32bb720-5a3a-4c11-877d-5273f289e5e8

chimmyten committed:
fix node height property not updating with content

chimmyten commented:
<img width="910" alt="Screenshot 2023-11-11 at 6 18 08 PM" src="https://github.com/JupyPod/JupyPod/assets/105993789/984852be-e0ef-4efb-8f70-bc21fbb752d8">
Not sure why, but the markdown pods have a black bar on the right side on my browser. This screen shot is after I checked out commit #[361952f]

chimmyten committed:
set the parent of node when dragged into a group
this is necessary for knowing which nodes to consider for auto-expanding
and shrinking;

chimmyten committed:
set data level when moving pod into group

chimmyten commented:
The above issue was resolved when I unplugged my wireless mouse. Very strange.

chimmyten committed:
fix inability to move scope into scope

chimmyten commented:
https://github.com/JupyPod/JupyPod/assets/105993789/a6898843-c918-4f24-bbcb-770242a28e3a

Screen recording on my laptop breaks some functionality for some reason.

chimmyten committed:
set parentNode attribute of nodes for multiple dragging

chimmyten commented:
https://github.com/JupyPod/JupyPod/assets/105993789/e54bb98c-8bfc-4562-8a09-55d67b799387

chimmyten committed:
nodeContextMenu and change param type of moveIntoScope

chimmyten committed:
create nodes directly in group

chimmyten commented:
https://github.com/JupyPod/JupyPod/assets/105993789/2bbf48d6-9183-41fd-8a4f-8df4a7cceda9

chimmyten commented:
https://github.com/JupyPod/JupyPod/assets/105993789/357247a8-4955-4f8d-8a0b-289c1c65af5c

This is the progress I've made so far with auto-resizing. As you can see, there are some bugs, and I also think there's an issue with moving pods into groups, which I need to go back and fix.

chimmyten committed:
fix auto-expand absolute pos issue

chimmyten committed:
auto layout groups recursively

chimmyten committed:
implement auto-shrinking

chimmyten commented:
https://github.com/JupyPod/JupyPod/assets/105993789/209151e5-4946-43df-98cd-63601ebf8522

chimmyten commented:
Ready for review
